internal/repository/repository.go

```diff
@@ -128,14 +128,25 @@ func (r *Repository) sortCachedPacks(blobs []restic.PackedBlob) []restic.PackedB
 // loadBlob tries to load and decrypt content identified by t and id from a
 // pack from the backend, the result is stored in plaintextBuf, which must be
 // large enough to hold the complete blob.
-func (r *Repository) loadBlob(ctx context.Context, id restic.ID, t restic.BlobType, plaintextBuf []byte) (int, error) {
-	debug.Log("load %v with id %v (buf len %v, cap %d)", t, id, len(plaintextBuf), cap(plaintextBuf))
+func (r *Repository) loadBlob(ctx context.Context, id restic.ID, t restic.BlobType, plaintextBuf []byte) ([]byte, int, error) {
+	if plaintextBuf == nil {
+		debug.Log("load %v with id %v (allocating buf)", t, id)
+	} else {
+		debug.Log("load %v with id %v (buf len %v, cap %d)", t, id, len(plaintextBuf), cap(plaintextBuf))
+	}
 
 	// lookup packs
 	blobs, found := r.idx.Lookup(id, t)
 	if !found {
 		debug.Log("id %v not found in index", id)
-		return 0, errors.Errorf("id %v not found in repository", id)
+		return plaintextBuf, 0, errors.Errorf("%v %v not found in repository", t, id)
+	}
+
+	if plaintextBuf == nil {
+		size := uint(restic.PlaintextLength(int(blobs[0].Length)))
+
+		debug.Log("size is %d, create buffer", size)
+		plaintextBuf = restic.NewBlobBuffer(int(size))
 	}
```

Comment:
Instead of treating nil as the special value, wouldn't it be more idiomatic to let the caller pass in arbitrary scratch space and allocate if it turns out to be too small? I can't recall a single instance where an empty []byte is treated differently from nil in the stdlib.

```
// untested
size := restic.PlaintextLength(int(blobs[0].Length))
if cap(plaintextBuf) > size {
	plaintextBuf = plaintextBuf[:size]
} else if cap(plaintextBuf) < size {
	debug.Log("size is %d, create buffer", size)
	plaintextBuf = restic.NewBlobBuffer(size)
}
```